aiida/tools/importexport/migration/v06_to_v07.py

```diff
@@ -0,0 +1,126 @@
+    `aiida.backends.djsite.db.migrations.00XX_<migration-name>.py`
+
+Where XX are the numbers in the migrations' documentation: REV. 1.0.XX
+And migration-name is the name of the particular migration.
+The individual SQLAlchemy database migrations may be found at:
+
+    `aiida.backends.sqlalchemy.migrations.versions.<id>_<migration-name>.py`
+
+Where id is a SQLA id and migration-name is the name of the particular migration.
+"""
+# pylint: disable=invalid-name
+from __future__ import division
+from __future__ import print_function
+from __future__ import absolute_import
+
+from aiida.tools.importexport.migration.utils import verify_metadata_version, update_metadata
+
+
+def migration_data_migration_legacy_process_attributes(data):
+    """Apply migration 0040 - REV. 1.0.40
+    Data migration for some legacy process attributes.
+
+    Attribute keys that are renamed:
+
+    * `_sealed` -> `sealed`
+
+    Attribute keys that are removed entirely:
+
+    * `_finished`
+    * `_failed`
+    * `_aborted`
+    * `_do_abort`
+
+    Finally, after these first migrations, any remaining process nodes are screened for the existence of the
+    `process_state` attribute. If they have it, it is checked whether the state is active or not, if not, the `sealed`
+    attribute is created and set to `True`.
+
+    :raises `~aiida.common.exceptions.IntegrityError`: if a Node, found to have attributes, cannot be found in the list
+        of exported entities.
+    :raises `~aiida.common.exceptions.IntegrityError`: if the 'sealed' attribute does not exist and the ProcessNode is
+        in an active state, i.e. `process_state` is one of ('created', 'running', 'waiting').
+        A log-file, listing all illegal ProcessNodes, will be produced in the current directory.
+    """
+    from aiida.common.exceptions import IntegrityError
+    from aiida.manage.database.integrity import write_database_integrity_violation
+
+    attrs_to_remove = ['_sealed', '_finished', '_failed', '_aborted', '_do_abort']
+    active_states = {'created', 'running', 'waiting'}
+    illegal_cases = []
+
+    for node_pk, content in data['node_attributes'].items():
+        try:
+            if data['export_data']['Node'][node_pk]['node_type'].startswith('process.'):
+                # Check if the ProcessNode has a 'process_state' attribute, and if it's non-active.
+                # Raise if the ProcessNode is in an active state, otherwise set `'sealed' = True`
+                process_state = content.get('process_state', '')
+                if process_state in active_states:
+                    # The ProcessNode is in an active state, and should therefore never have been allowed
+                    # to be exported. The Node will be added to a log that is saved in the working directory,
+                    # then an IntregityError will be raised, since the archive needs to be migrated manually.
+                    uuid_pk = data['export_data']['Node'][node_pk].get('uuid', node_pk)
```

Comment:
Do we really have cases where there is no `uuid` and be safe we need to use the `node_pk`?

Reply:
Hopefully not, but for what is supposed to be a helpful log for us/the user, I'd rather try and catch the ball as many times as possible, so to say. However, if the identifier was used in the actual code, it should definitely be more strict, i.e. only use either UUID or PK.